src/core/server/audit_trail/types.ts

```diff
@@ -19,17 +19,78 @@
 import { KibanaRequest } from '../http';
 
 /**
- * Event to audit.
+ * Audit event using ECS format: https://www.elastic.co/guide/en/ecs/1.5/index.html
  * @public
  *
  * @remarks
  * Not a complete interface.
  */
-export interface AuditableEvent {
+export interface AuditEvent {
   message: string;
-  type: string;
+  event: {
+    action: string;
+    category?: EventCategory | EventCategory[];
+    type?: EventType | EventType[];
+    outcome?: EventOutcome;
+    module?: string;
+    dataset?: string;
+  };
+  user: {
+    name: string;
+    email?: string;
+    full_name?: string;
+    hash?: string;
+  };
+  kibana?: {
+    session_id?: string;
+    user_roles?: string[];
+    saved_objects?: Array<{
+      id?: string;
+      type: string;
+      namespaces?: string[];
+    }>;
```

Comment:
1. 

Agreed, I am already extracting the `id` field from the saved object. However, when the create operation fails no `id` gets generated so if the `id` field would be required I wouldn't be able to log `type` and `namespaces` info with the event. 

I don't know how strict we need to be with the schema here and what value a tighter schema adds. ECS is quite a loose standard so would fit fairly well to add as much information as necessary / useful but not have too many required fields. 

Happy to make this tighter though if we think that would benefit auditors. 

2.

Yeh, I think you're raising a good point here. I did one iteration where I just captured the current space as `kibana.namespace` but then saw the add to / remove from namespaces methods so felt like this approach would simplify capturing those events. More than happy to change this though

Reply:
> Agreed, I am already extracting the `id` field from the saved object. However, when the create operation fails no `id` gets generated so if the `id` field would be required I wouldn't be able to log `type` and `namespaces` info with the event.
Actually, the ID is generated by Kibana before the Elasticsearch call is made, so there is never a "no ID" case when using the Saved Objects Client (and the Elasticsearch audit trail will reflect the ID as well).

> I don't know how strict we need to be with the schema here and what value a tighter schema adds. ECS is quite a loose standard so would fit fairly well to add as much information as necessary / useful but not have too many required fields.
>    Happy to make this tighter though if we think that would benefit auditors.

Well I think keeping the schema as tight as possible helps ensure that the type checker catches oversights in our implementation, that's all.

Reply:
> Agreed, I am already extracting the `id` field from the saved object.

Sorry, I forgot to add in my comment above:

Your current implementation does not do this. For the `create` operation, you're extracting the `options.id` field (which is optional) -- you'll need to change it to use the ID from the result instead 🙂 Same goes for the `bulkCreate` operation.

Reply:
> > Agreed, I am already extracting the `id` field from the saved object. However, when the create operation fails no `id` gets generated so if the `id` field would be required I wouldn't be able to log `type` and `namespaces` info with the event.
> 
> Actually, the ID is generated by Kibana before the Elasticsearch call is made, so there is never a "no ID" case when using the Saved Objects Client (and the Elasticsearch audit trail will reflect the ID as well).

While Kibana does generate the ID, I don't think it's ever returned up to the wrappers / out of the client in failure scenarios. If that's the case (and if it's not something we can realistically change), then we can't record the ID as part of this failed create attempt.

Reply:
> While Kibana does generate the ID, I don't think it's ever returned up to the wrappers / out of the client in failure scenarios. If that's the case (and if it's not something we can realistically change), then we can't record the ID as part of this failed create attempt.

The ID isn't returned for `create` failures, but it is for `bulkCreate` failures.

<details>
<summary><strong>create</strong> example conflict (HTTP 409 response code)</summary>

```json
{
    "statusCode": 409,
    "error": "Conflict",
    "message": "[foo:e73d9c00-e0ad-11ea-9eb5-e1453e1277e6]: version conflict, required seqNo [27], primary term [1]. current document has seqNo [28] and primary term [1]: version_conflict_engine_exception"
}
```
</details>

<details>
<summary><strong>bulkCreate</strong> example conflict (HTTP 200 response code)</summary>

```json
{
    "saved_objects": [
        {
            "id": "e73d9c00-e0ad-11ea-9eb5-e1453e1277e6",
            "type": "foo",
            "error": {
                "statusCode": 409,
                "error": "Conflict",
                "message": "Saved object [foo/e73d9c00-e0ad-11ea-9eb5-e1453e1277e6] conflict"
            }
        }
    ]
}
```
</details>

That said, we should never see version conflicts for create operations where the consumer did not specify the object ID. A quick peek at the [Elasticsearch API tests](https://github.com/elastic/elasticsearch/blob/3bce3ec9c6f5a3106f26352f01270aec365b9b54/client/rest-high-level/src/test/java/org/elasticsearch/client/CrudIT.java#L456) indicate there aren't any other 4xx create error cases we should expect to encounter.

However, we should still attempt to extract the object ID from the Saved Objects Client response, because it will be returned in the success case for both `create` and `bulkCreate`.